networkx/algorithms/polynomials/tutte.py

```diff
@@ -0,0 +1,180 @@
+    connected components of `A` [1]_:
+
+    .. math::
+
+        T_G(x, y) = \sum_{A \in E} (x-1)^{c(A) - c(G)} (y-1)^{c(A) + |A| - n(G)}
+
+    Def 2 (spanning tree expansion): For `G` an undirected graph, `T` a spanning
+    tree of `G`, `i(T)` the internal activity of `T`, and `e(T)` the external
+    activity of `T` [2]_:
+
+    .. math::
+
+        T_G(x, y) = \sum_{T \text{ a spanning tree of } G} x^{i(T)} y^{e(T)}
+
+    Def 3 (deletion-contraction recurrence): For `G` an undirected graph, `G-e`
+    the graph obtained from `G` by deleting edge `e`, `G/e` the graph obtained
+    from `G` by contracting edge `e`, `k(G)` the number of cut-edges of `G`,
+    and `l(G)` the number of loops of `G`:
+
+    .. math::
+        T_G(x, y) = \begin{cases}
+    	   x^{k(G)} y^{l(G)}, & \text{if all edges are cut-edges or loops} \\
+           T_{G-e}(x, y) + T_{G/e}(x, y), & \text{otherwise, for $e$ not a cut-edge or loop}
+        \end{cases}
+
+    Parameters
+    ----------
+    G : NetworkX graph
+
+    simplify : bool
+        determines if the returned expression will be in simplified form
+
+    Returns
+    -------
+    instance of :class:`sympy.core.add.Add`
+        A Sympy expression representing the Tutte polynomial for `G`.
+
+    Examples
+    --------
+    >>> C = nx.cycle_graph(5)
+    >>> nx.tutte_polynomial(C)
+    x**4 + x**3 + x**2 + x + y
+
+    >>> D = nx.diamond_graph()
+    >>> nx.tutte_polynomial(D)
+    x**3 + 2*x**2 + 2*x*y + x + y**2 + y
+
+    Notes
+    -----
+    This function implements the deletion-contraction recurrence (Definition 3,
+    described above). Edge contraction is defined and deletion-contraction is
+    introduced in [3]_. Combinatorial meaning of the coefficients is introduced
+    in [4]_. Universality, properties, and applications are discussed in [5]_.
+
+    See Also
+    --------
+    _get_cut_edges
+
+    References
+    ----------
+    .. [1] Y. Shi, M. Dehmer, X. Li, I. Gutman,
+       "Graph Polynomials," p. 14
+    .. [2] Y. Shi, M. Dehmer, X. Li, I. Gutman,
+       "Graph Polynomials," p. 46
+    .. [3] D. B. West,
+       "Introduction to Graph Theory," p. 84
+    .. [4] G. Coutinho,
+       "A brief introduction to the Tutte polynomial"
+       Structural Analysis of Complex Networks, 2011
+       https://homepages.dcc.ufmg.br/~gabriel/seminars/coutinho_tuttepolynomial_seminar.pdf
+    .. [5] J. A. Ellis-Monaghan, C. Merino,
+       "Graph polynomials and their applications I: The Tutte polynomial"
+       Structural Analysis of Complex Networks, 2011
+       https://arxiv.org/pdf/0803.3079.pdf
+    """
+    import sympy
+
+    G = nx.MultiGraph(G)
+
+    components = list(nx.connected_components(G))
+    if len(components) > 1:
+        t = 1
+        for component in components:
+            t *= tutte_polynomial(G.subgraph(component), False)
+        return sympy.simplify(t)
+
+    cut_edges = _get_cut_edges(G)
+    loops = [e for e in G.edges if e[0] == e[1]]
```

Comment:
Use `loops = set(nx.selfloop_edges(G))`  
The `set` choice is because in the next line we are checking if `i not in loop`.

Reply:
I'm happy to make a change in this direction, but I think `loops = list(nx.selfloop_edges(G, keys=True))` would be more appropriate, no? I may have multiple loops on the same node, and that is important for `sympy.Symbol("y") ** len(loops)`